projecttacoma/bulk-export-app
# TACOSTRAT 1528 Bulk Export Server URL Input

# Summary

This PR adds support for inputing the url of the bulk export server on the home page of the application so that it can use a server other than the one that is specified in the environment variables of the application.


[JIRA Issue](https://jira.mitre.org/browse/TACOSTRAT-1528)

## New behavior

The home page of the application at `http://localhost:3000` now has an input field that takes in a url for the location of the bulk export server. The placeholder value for this field comes from the environment variables, so that for testing and development of the app, I didn't have to keep entering the server url. This input field takes in a string, validates that it is a url for the location of the bulk export server by using the bulk export server's `/metadata` endpoint. If reaching out to this endpoint fails, the input field will turn red and a notification should tell the user that a bad url was provided. Extra validation was added to look into the returned metadata response. The app looks for there to be an exposed `$export` operation in the metadata compatibility statement. If this isn't found, the user will be warned but it will not result in an error, in case in the future the compatibility statement changes form.

The validation of the input was done using [Mantine use-field hook](https://mantine.dev/form/use-field/) which allows for easy asynchronous form validation on a single field form.

When the user enters a valid url, the url they entered is pushed to the browser's local storage and they are navigated to the next page. That way when the user refreshes subsequent pages, the bulk export server url persists. To accomplish this and integrate it with the app's recoil state management, I used an npm package [recoil-persist](https://www.npmjs.com/package/recoil-persist).

In order to use the browser storage like this, I had to create a new file called `providers.tsx` which is a wrapper component that wraps its children in the `RecoilRoot` provider. I then imported this component inside the root `layout.tsx` file using next.js [Lazy Loading](https://nextjs.org/docs/app/building-your-application/optimizing/lazy-loading#skipping-ssr) `{ssr : false}` to disable ssr which prevents recoil and the recoil-presist package from trying to initialize state values server side where the browser local storage does not exist.

## Code changes
Header changes

changed to use the same header and styles as the other pages.
- `src/app/export-execution/layout.tsx`
- `src/app/query-selector/layout.tsx`
- `src/app/query-builder/layout.tsx`


Other Changes
- `src/app/export-execution/page.tsx` - made the grid more responsive by changing to vertical alignment after a certain breakpoint.
- `src/app/global.module.css` - changed the app header and simplified the styles.
- `src/app/layout.tsx` changed to use the providers component and lazy loading as described above.
- `src/app/page.module.css` - deleted and moved styles into the global styles css file.
- `src/app/page.tsx` - added the input for bulk server url and form validation logic.
- `src/app/providers.tsx` - added to fix the issue described above.
- `src/app/query-selector/page.tsx` - changed the server url to come from the recoil state rather than environment variables.
- `src/components/export-execution/file-downloads.tsx` - deleted because it was not being used. Must have stayed from when I changed the name of the file in a previous PR.
- `src/components/export-execution/resource-file-preview.tsx` - changed a style name that I used on the home page.
- `src/components/header.tsx` - added reusable header component to be consistent across all pages of the app. Also now included the url of the bulk-export-server, mostly for testing purposes so that I could verify that the recoil store and local storage was working as expected with refreshes of the pages.
- `src/components/query-selector/export-type.tsx` - changed the radius of the input to be consistent with the home page.
- `src/state/bulk-server-url-state.ts` - recoil state for the bulk export url that uses the recoil-persist npm package.
- `src/util/exportRequestBuilders.ts` - changed the server url to come from the recoil state rather than environment variables.
- `src/util/validators.ts` - new file for function that validate url paths. Currently only has the url path validation for the bulk export server url functionality that was added in this PR.
 
# Testing guidance
- Run the bulk export server
- `npm run check`
- `npm run build`
- `npm run start`
- Enter `http://localhost:3000` into the url field and verify that the success notification comes up and you are navigated to the next page.
- Look in the browser's local storage and you should see a value from recoil-persist for the `bulkExportServerURLState`. Verify that this matches.
- Refresh page and verify the url persists.
- Then try a bad url, some random website or an ill formed url will work. Verify that an error notification comes up and that you are not navigated to the next page.
